tests/test_dragon_installer.py

```diff
@@ -0,0 +1,519 @@
+
+                asset1 = GitReleaseAsset(requester, headers, attributes, completed)
+
+                asset1_name = mock_archive_name
+                asset1_name = asset1_name.replace("py3.9", python_version)
+                asset1_name = asset1_name.replace("dragon-0.8", dragon_version)
+                asset1_name = asset1_name.replace("CRAYEX-", platform)
+
+                setattr(
+                    asset1,
+                    "_browser_download_url",
+                    _git_attr(value=f"http://foo/{asset1_name}"),
+                )
+                setattr(asset1, "_name", _git_attr(value=asset1_name))
+                assets[asset1_name] = asset1
+
+    return assets
+
+
+def test_cleanup_no_op(test_dir: str) -> None:
+    """Ensure that the cleanup method doesn't bomb when called with
+    missing archive path and extraction directory; simulate a failed
+    download"""
+    archive_path = pathlib.Path(test_dir) / mock_archive_name
+    extract_path = pathlib.Path(test_dir) / mock_archive_name.replace(".tar.gz", "")
+
+    # confirm assets do not exist
+    assert not archive_path.exists()
+    assert not extract_path.exists()
+
+    # call cleanup. any exceptions should break test...
+    cleanup(archive_path, extract_path)
+
+
+def test_cleanup_empty_extraction_directory(test_dir: str) -> None:
+    """Ensure that the cleanup method works when the extraction directory
+    is empty"""
+    archive_path = pathlib.Path(test_dir) / mock_archive_name
+    extract_path = pathlib.Path(test_dir) / mock_archive_name.replace(".tar.gz", "")
+    extract_path.mkdir()
+
+    # verify archive doesn't exist & folder does
+    assert not archive_path.exists()
+    assert extract_path.exists()
+
+    cleanup(archive_path, extract_path)
+
+    # verify folder is gone after cleanup
+    assert not archive_path.exists()
+    assert not extract_path.exists()
+
+
+def test_cleanup_nonempty_extraction_directory(test_dir: str) -> None:
+    """Ensure that the cleanup method works when the extraction directory
+    is NOT empty"""
+    archive_path = pathlib.Path(test_dir) / mock_archive_name
+    extract_path = pathlib.Path(test_dir) / mock_archive_name.replace(".tar.gz", "")
+    extract_path.mkdir()
+
+    num_files = 10
+    for i in range(num_files):
+        content = extract_path / f"{i:04}.txt"
+        content.write_text(f"i am file {i}\n")
+
+    files = list(extract_path.rglob("*.txt"))
+
+    assert len(files) == num_files
+
+    cleanup(archive_path, extract_path)
+
+    # verify folder is gone after cleanup
+    assert not archive_path.exists()
+    assert not extract_path.exists()
+
+
+def test_cleanup_no_extract_path(test_dir: str) -> None:
+    """Ensure that the cleanup method doesn't bomb when called with
+    missing extraction directory; simulate failed extract"""
+    archive_path = pathlib.Path(test_dir) / mock_archive_name
+    extract_path = pathlib.Path(test_dir) / mock_archive_name.replace(".tar.gz", "")
```

Comment:
Looks like these two are recalculated  for multiple tests, consider making them local fixtures

Reply:
Done!